guide.md

```diff
@@ -0,0 +1,350 @@
+
+# download rules_nixpkgs
+http_archive(
+    name = "io_tweag_rules_nixpkgs",
+    strip_prefix = "rules_nixpkgs-9f08fb2322050991dead17c8d10d453650cf92b7",
+    urls = ["https://github.com/tweag/rules_nixpkgs/archive/9f08fb2322050991dead17c8d10d453650cf92b7.tar.gz"],
+    sha256 = "46aa0ca80b77848492aa1564e9201de9ed79588ca1284f8a4f76deb7a0eeccb9",
+)
+
+# load everything that rules_nixpkgs rules need to work
+load("@io_tweag_rules_nixpkgs//nixpkgs:repositories.bzl", "rules_nixpkgs_dependencies")
+rules_nixpkgs_dependencies()
+```
+
+Now we can select a version of the Nixpkgs repository from which all of our
+packages will be drawn. Any Git commit hash or tag is allowed. This way, we
+select a consistent set of package revisions in one go instead of having to
+maintain dependency versions for individual dependencies. Often this will be a
+tag for a stable release, which ensures that there has been especially
+stringent verification that packages in it will function correctly individually
+and play well together. Here we use the 22.05 release:
+
+```bazel
+load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_git_repository")
+nixpkgs_git_repository(
+    name = "nixpkgs",
+    revision = "22.05",
+    sha256 = "",
+)
+```
+
+# A hermetic CC toolchain
+
+We can now define a CC toolchain that will be provisioned by Nix using
+package definitions from the 22.05 release of Nixpkgs.
+
+```bazel
+load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_cc_configure")
+nixpkgs_cc_configure(
+  repository = "@nixpkgs",
+  name = "nixpkgs_config_cc",
+)
+```
+
+This will select the default `stdenv.cc` package as the compiler, drawing
+auxiliary tools from `stdenv.cc.bintools`. However, one can write a little bit of
+Nix to configure this:
+
+```bazel
+load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_cc_configure")
+nixpkgs_cc_configure(
+  repository = "@nixpkgs",
+  # nix_file_content = "(import <nixpkgs> {}).gcc11",
+  attribute_path = "gcc11",
+  name = "nixpkgs_config_cc",
+)
+```
+
+`(import <nixpkgs> {})` loads the entire package set into an attribute set
+(Nix's equivalent of a record), and `.gcc11` selects one value from it. It is
+possible to customize the compiler further using more Nix if desired: you could,
+for instance, point Nix to a custom fork of Clang containing support for a new
+architecture, and have everything work seamlessly.
+
+With this, we now have a fully hermetic CC toolchain that will behave correctly
+across platforms. It remains to tell Bazel to use it by default, instead of
+`@local_config_cc`. To do so, we put the following into `.bazelrc` at the root
+of the repository:
+
+```
+build --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+build --crosstool_top=@nixpkgs_config_cc//:toolchain
+```
+
+Now `bazel build` and `bazel run` should use this toolchain by default for
+C/C++ compilation:
+```
+> bazel build //src:hello-world
```

Comment:
This snippet does not "prove" that a different toolchain is indeed used. Perhaps building with `--toolchain_resolution_debug` could illustrate the switch better?

Reply:
Added a note